ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java

```diff
@@ -331,6 +334,94 @@ public Void visitSubscriptExpression(
       return null;
     }
 
+    @Override
+    public Void visitCreateArrayExpression(
+        final CreateArrayExpression exp,
+        final ExpressionTypeContext context
+    ) {
+      if (exp.getValues().isEmpty()) {
+        throw new KsqlException(
+            "Array constructor cannot be empty. Please supply at least one element "
+                + "(see https://github.com/confluentinc/ksql/issues/4239).");
+      }
+
+      final List<SqlType> sqlTypes = exp
+          .getValues()
+          .stream()
+          .map(val -> {
+            process(val, context);
+            return context.getSqlType();
+          })
+          .filter(Objects::nonNull)
+          .collect(Collectors.toList());
+
+      if (sqlTypes.size() == 0) {
+        throw new KsqlException("Cannot construct an array with all NULL elements "
+            + "(see https://github.com/confluentinc/ksql/issues/4239). As a workaround, you may "
+            + "cast the NULL value to a desired type.");
```

Comment:
nit: consider changing 'the NULL' to 'a NULL'.  There could be multiple nulls and only one needs to be none-null